framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java

```diff
@@ -1066,178 +895,79 @@ public void typeProcess(TypeElement e, TreePath p) {
-    private final String suppressWarningsClassName = SuppressWarnings.class.getCanonicalName();
     /**
-     * Finds the tree that is a {@code @SuppressWarnings} annotation.
-     *
-     * @param tree a class, method, or variable tree annotated with {@code @SuppressWarnings}
-     * @return tree for {@code @SuppressWarnings} or {@code default} if one isn't found
+     * Method that gets called exactly once at shutdown time of the JVM. Checkers can override this
+     * method to customize the behavior.
      */
-    private Tree findSuppressWarningsTree(Tree tree) {
-        List<? extends AnnotationTree> annotations;
-        if (TreeUtils.isClassTree(tree)) {
-            annotations = ((ClassTree) tree).getModifiers().getAnnotations();
-        } else if (tree.getKind() == Tree.Kind.METHOD) {
-            annotations = ((MethodTree) tree).getModifiers().getAnnotations();
-        } else {
-            annotations = ((VariableTree) tree).getModifiers().getAnnotations();
+    protected void shutdownHook() {
+        if (hasOption("resourceStats")) {
+            // Check for the "resourceStats" option and don't call shouldAddShutdownHook
+            // to allow subclasses to override shouldXXX and shutdownHook and simply
+            // call the super implementations.
+            printStats();
         }
+    }
 
-        for (AnnotationTree annotationTree : annotations) {
-            if (AnnotationUtils.areSameByName(
-                    TreeUtils.annotationFromAnnotationTree(annotationTree),
-                    suppressWarningsClassName)) {
-                return annotationTree;
-            }
+    /** Print resource usage statistics. */
+    protected void printStats() {
+        List<MemoryPoolMXBean> memoryPools = ManagementFactory.getMemoryPoolMXBeans();
+        for (MemoryPoolMXBean memoryPool : memoryPools) {
+            System.out.println("Memory pool " + memoryPool.getName() + " statistics");
+            System.out.println("  Pool type: " + memoryPool.getType());
+            System.out.println("  Peak usage: " + memoryPool.getPeakUsage());
         }
-        throw new BugInCF("Did not find @SuppressWarnings: " + tree);
     }
 
-    private BugInCF wrapThrowableAsBugInCF(String where, Throwable t, @Nullable TreePath p) {
-        return new BugInCF(
-                where
-                        + ": unexpected Throwable ("
-                        + t.getClass().getSimpleName()
-                        + ")"
-                        + ((p == null)
-                                ? ""
-                                : " while processing "
-                                        + p.getCompilationUnit().getSourceFile().getName())
-                        + (t.getMessage() == null ? "" : "; message: " + t.getMessage()),
-                t);
-    }
+    ///////////////////////////////////////////////////////////////////////////
+    /// Reporting type-checking errors; the main entry point is report()
```

Comment:
I'm not sure what the "main entry point" is supposed to mean. Can you clarify?
The two methods that visitors should use are `report()` and `message(Diagnostic.Kind kind, String msg, Object... args)`, so maybe list those two?

Reply:
`message` is documented as being "only for exceptional cases".  So `report` is the method that clients should prefer.